core/src/apps/solana/token_account.py

```diff
@@ -0,0 +1,65 @@
+from typing import TYPE_CHECKING
+
+from trezor.crypto import base58
+
+if TYPE_CHECKING:
+    from trezor.messages import SolanaTxTokenAccountInfo
+
+ASSOCIATED_TOKEN_ACCOUNT_PROGRAM = "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
+
+SEED_CONSTANT = "ProgramDerivedAddress"
+
+
+def assert_is_associated_token_account(
+    base_address: bytes,
+    token_account_address: bytes,
+    token_program: bytes,
+    token_mint: bytes,
+) -> None:
+    from trezor.crypto.hashlib import sha256
+
+    # based on the following sources:
+    # https://spl.solana.com/associated-token-account#finding-the-associated-token-account-address
+    # https://github.com/solana-labs/solana-web3.js/blob/745437f79bf4e268929e7740cfc517a2fcee980f/packages/library-legacy/src/publickey.ts#L212
+    for seed_bump in range(255, 0, -1):
+        seed = (
+            base_address
+            + token_program
+            + token_mint
+            + bytes([seed_bump])
+            + base58.decode(ASSOCIATED_TOKEN_ACCOUNT_PROGRAM)
+            + SEED_CONSTANT.encode("utf-8")
+        )
+
+        account = sha256(seed).digest()
+
+        if account == token_account_address:
```

Comment:
do we have to check this way? what is the blockchain-side condition that determines the seed?
if there are multiple valid seeds, and resulting associated accounts (for multiple seeds), this allows an attacker to supply an unintended one. it's a small problem if the ownership is still for the base address, but it _is_ a problem nevertheless

Reply:
[This is the function](https://github.com/solana-labs/solana/blob/8fbe033eaca693ed8c3e90b19bc3f61b32885e5e/sdk/program/src/pubkey.rs#L495) which iterates the seed. And deeper into [syscalls::sol_try_find_program_address](https://github.com/solana-labs/solana/blob/8fbe033eaca693ed8c3e90b19bc3f61b32885e5e/programs/bpf_loader/src/syscalls/mod.rs#L769.).

Reply:
so in other words we can check if the generated `account` is outside the curve. our cryptography library can do it but it's not exposed :(

still, it should not be possible to call ATAP Create with an associated address that is not the first one found by this algorithm, correct?

Reply:
> so in other words we can check if the generated account is outside the curve. our cryptography library can do it but it's not exposed :(

We could check ourselves indeed.

> still, it should not be possible to call ATAP Create with an associated address that is not the first one found by this algorithm, correct?

Correct.